### PR TITLE
1415 promise events

### PR DESCRIPTION
**Pull request by boutell**
Implements #1415 

Adds one feature: a shortcut to add a method while registering it as an event handler. 

See the tests for examples of use.

### CODE REVIEW

**Comment by boutell** [comment]
Assigned to the enterprise team because this can't be one I just run by a single developer who doesn't spend time in the back end of Apostrophe.

File: lib/modules/apostrophe-module/index.js [hunk 1]
```diff
@@ -681,5 +681,87 @@ module.exports = {
       });
     }
 
+    self.eventHandlers = {};
```
Comment (austinstarin) [inline lib/modules/apostrophe-module/index.js]: dead code?

File: lib/modules/apostrophe-module/index.js [hunk 2]
```diff
@@ -681,5 +681,87 @@ module.exports = {
...
+    // "Why can't I use a methodName that is identical to
+    // the event name?" Doing so sets you up for an accidental
+    // collision with other event handlers in subclasses.
+    // Your method name should describe what your method does
+    // in response to the event.
+
+    self.on = function(name, methodName, fn) {
+      var moduleName, eventName, index;
+      index = name.indexOf(':');
+      if (_.camelCase(name) === _.camelCase(methodName)) {
```
Comment (austinstarin) [inline lib/modules/apostrophe-module/index.js]: Test needs to happen on name not module part.

File: lib/modules/apostrophe-module/index.js [hunk 3]
```diff
@@ -681,5 +681,87 @@ module.exports = {
...
+        self[methodName] = fn;
+      }
+      eh[eventName] = eh[eventName] || [];
+      if (_.find(eh[eventName], function(item) {
+        return (item.moduleName === self.__meta.name) && (item.methodName === methodName);
+      })) {
+        // The "event name and method name must differ" rule helps
+        // prevent this situation, but it is still possible and a
+        // warning is in order
+        self.apos.utils.warn('The method ' + methodName + ' has been registered twice for the event\n' + eventName + ' in the module ' + self.__meta.name + '. It will only be called once.\n\nYou should rename the method in your code to differ from that in the module\nit extends so they both run.\n\nIf you are intentionally overriding it, assign a new function to\nself.' + methodName + ' instead.');
```
Comment (austinstarin) [inline lib/modules/apostrophe-module/index.js]: Would not be invoked once, but twice.

**Comment by austinstarin** [comment]
Should we require innovations of this? You need to be able to do `pageBeforeSend` in this new style at the very least...

**Comment by boutell** [comment]
@austinstarin please see the new commit in which I address the concerns raised in the review, including implementing actual emitting of the events for all of the current callAll situations and a unit test demonstrating that you can listen to those.

**Comment by boutell** [comment]
Example from the tests:

```
      modules: {
        'test1': {
          alias: 'test1',
          construct: function(self, options) {
            self.on('apostrophe-docs:beforeInsert', 'beforeInsertReverseTitle', function(req, doc, options) {
              if (doc.type === 'default') {
                return Promise.delay(50).then(function() {
                  doc.title = doc.title.split('').reverse().join('');
                });
              }
            });
            self.on('apostrophe:modulesReady', 'modulesReadyCoreEventsWork', function() {
              coreEventsWork = true;
            });
            // Old school callAll works too (for now)
            self.docBeforeInsert = function(req, doc, options, callback) {
              doc.oldSchool = true;
              return setImmediate(callback);
            };
          }
        },
      }
```

Use of Promise.delay is a contrived way of proving you can do async stuff and it’ll wait for you before continuing.

Note the `apostrophe:modulesReady` example. A few callAlls were coming from the Apostrophe object itself. After staring at that for a bit, I concluded that the `apos` object simply needs to be capable of emitting promise events too, and its namespace is of course just `apostrophe`.

Also, a thought experiment: if I teach `apostrophe-pieces` at project level to emit an event, and I listen for that event, I won't "hear" it unless I listen like this...

```
// one of my piece types
apos.on('people:coolEvent', ...)
// another one
apos.on('trucks:coolEvent', ...)
```

Etc.

I think this is fine because sometimes people extend modules they also use directly, and that would lead to confusion and no way to tell the events apart if our handlers also fired for parent classes.

However, nothing stops us from introducing a syntax for that later if it's desired.

**Pull request merged**